File: frontend/app_flowy/lib/workspace/application/grid/cell/cell_service/data_persistence.dart
part of 'cell_service.dart';

abstract class _GridCellDataPersistence<D> {
  Future<Option<FlowyError>> save(D data);
}

class CellDataPersistence implements _GridCellDataPersistence<String> {
  final GridCell gridCell;

  CellDataPersistence({
    required this.gridCell,
  });
  final CellService _cellService = CellService();

  @override
  Future<Option<FlowyError>> save(String data) async {
    final fut = _cellService.updateCell(
      gridId: gridCell.gridId,
      fieldId: gridCell.field.id,
      rowId: gridCell.rowId,
      data: data,
    );

    return fut.then((result) {
      return result.fold(
        (l) => none(),
        (err) => Some(err),
      );
    });
  }
}

@freezed
class CalendarData with _$CalendarData {
  const factory CalendarData({required DateTime date, String? time}) = _CalendarData;
}

class DateCellDataPersistence implements _GridCellDataPersistence<CalendarData> {
  final GridCell gridCell;
  DateCellDataPersistence({
    required this.gridCell,
  });

  @override
  Future<Option<FlowyError>> save(CalendarData data) {
    var payload = DateChangesetPayload.create()..cellIdentifier = _cellIdentifier(gridCell);

    final date = (data.date.millisecondsSinceEpoch ~/ 1000).toString();
    payload.date = date;

    if (data.time != null) {
      payload.time = data.time!;
    }

    return GridEventUpdateDateCell(payload).send().then((result) {
      return result.fold(
        (l) => none(),
        (err) => Some(err),
      );
    });
  }
}

CellIdentifierPayload _cellIdentifier(GridCell gridCell) {
  return CellIdentifierPayload.create()
    ..gridId = gridCell.gridId
    ..fieldId = gridCell.field.id
    ..rowId = gridCell.rowId;
}

File: frontend/appflowy_tauri/src/appflowy_app/stores/effects/database/cell/data_persistence.ts
import { Result } from 'ts-results';
import { CellBackendService, CellIdentifier } from './cell_bd_svc';
import { CalendarData } from './controller_builder';
import { DateChangesetPB, FlowyError, CellIdPB } from '@/services/backend';
import { DatabaseEventUpdateDateCell } from '@/services/backend/events/flowy-database';

export abstract class CellDataPersistence<D> {
  abstract save(data: D): Promise<Result<void, FlowyError>>;
}

export class TextCellDataPersistence extends CellDataPersistence<string> {
  constructor(public readonly cellId: CellIdentifier) {
    super();
  }

  save(data: string): Promise<Result<void, FlowyError>> {
    return CellBackendService.updateCell(this.cellId, data);
  }
}

export class DateCellDataPersistence extends CellDataPersistence<CalendarData> {
  constructor(public readonly cellIdentifier: CellIdentifier) {
    super();
  }

  save(data: CalendarData): Promise<Result<void, FlowyError>> {
    const payload = DateChangesetPB.fromObject({ cell_path: _makeCellPath(this.cellIdentifier) });
    payload.date = ((data.date.getTime() / 1000) | 0).toString();
    payload.is_utc = true;
    if (data.time !== undefined) {
      payload.time = data.time;
    }
    payload.include_time = data.includeTime;
    return DatabaseEventUpdateDateCell(payload);
  }
}

function _makeCellPath(cellIdentifier: CellIdentifier): CellIdPB {
  return CellIdPB.fromObject({
    view_id: cellIdentifier.viewId,
    field_id: cellIdentifier.fieldId,
    row_id: cellIdentifier.rowId,
  });
}
